sds/src/scanner/mod.rs

```diff
@@ -44,30 +45,107 @@ where
     }
 }
 
-pub trait CompiledRuleTrait: Send + Sync {
+pub trait CompiledRuleDyn: Send + Sync {
     fn get_match_action(&self) -> &MatchAction;
     fn get_scope(&self) -> &Scope;
+
+    #[allow(clippy::too_many_arguments)]
+    fn get_string_matches(
+        &self,
+        content: &str,
+        path: &Path,
+        caches: &mut CachePoolGuard<'_>,
+        group_data: &mut AHashMap<TypeId, Box<dyn Any>>,
+        exclusion_check: &ExclusionCheck<'_>,
+        excluded_matches: &mut AHashSet<String>,
+        match_emitter: &mut dyn MatchEmitter,
+        should_keywords_match_event_paths: bool,
+    );
+}
+
+pub trait CompiledRule: Send + Sync {
+    /// Data that is instantiated once per string being scanned, and shared with all rules that
+    /// have the same `GroupData` type. `Default` will be used to initialize this data.
+    type GroupData: Default + 'static;
```

Comment:
I'm not clear on the motivation of this change and on reasons why it is needed. Would it be possible to add this `GroupData` only in the ML rule implementing `CompiledRuleDyn`?

Reply:
This is a rust trick that @fuchsnj did to be able to allocate a GroupData only for the CompiledRule that has a GroupData type associated with it.
For Regex
impl CompiledRule for RegexCompiledRule {
    // no special data
    type GroupData = ();
    ...
    }
    
    
    So we don't add entry to hashmap for regex and it does not introduce any performance penalty
    
    The CompiledRuleDyn is a trick to downcast the the right groupData to its specific structure and call the CompiledRule specific implementation

Reply:
Ok it makes more sense now. Would it be possible to add that in the rustdoc of `CompiledRuleDyn` and `CompiledRule`? I think it will be easier to understand the motivation of this structure for future readers. (Usually I think code is self-explained but this time, I believe some comments would really help [because the code does some "convolutions" to find its way])

Reply:
sure let me do that